src/udiskslinuxfilesystemhelpers.c

```diff
@@ -0,0 +1,181 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Author: Vratislav Podzimek <vpodzime@redhat.com>
+ *
+ */
+
+#include <glib.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <blockdev/fs.h>
+
+#include "udiskslinuxfilesystemhelpers.h"
+#include "udiskslogging.h"
+
+
+static gboolean recursive_chown (const gchar *directory, uid_t caller_uid, gid_t caller_gid)
+{
+  GDir * gdir = NULL;
+  const gchar *fname = NULL;
+  g_autofree gchar *path = NULL;
+  GError *local_error = NULL;
+
+  gdir = g_dir_open (directory, 0, &local_error);
+  if (gdir == NULL)
+    {
+      g_clear_error (&local_error);
+      return FALSE;
+    }
+
+  if (chown (directory, caller_uid, caller_gid) != 0)
+    {
+      g_dir_close (gdir);
+      return FALSE;
+    }
+
+  while ((fname = g_dir_read_name (gdir)))
+    {
+      path = g_strdup_printf ("%s/%s", directory, fname);
```

Comment:
I still don't like this memory allocation and deallocation for every file, but we can live with it, I guess.